### PR TITLE
ev3dev/fb: add support for XRGB

### PR DESCRIPTION
**Pull request by kortschak**
@dlech Please take a look.

This will provide the construction function required for https://github.com/ev3go/ev3/pull/3.

### CODE REVIEW

File: fb/xrgb.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+	"errors"
+	"image"
+	"image/color"
+	"image/draw"
+)
+
+// NewXRGB returns a new XRGB image with the given bounds.
+func NewXRGB(r image.Rectangle) *XRGB {
+	w, h := r.Dx(), r.Dy()
+	stride := 2 * w
```
Comment (dlech) [inline fb/xrgb.go]: Should this be 4 instead of 2?

File: fb/xrgb.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+	if stride == 0 {
+		stride = 4 * w
+	}
+	if len(pix) < stride*h {
+		return nil, errors.New("ev3dev: bad pixel buffer length")
+	}
+	return &XRGB{Pix: pix, Stride: stride, Rect: r}, nil
+}
+
+// XRGB is an in-memory image whose At method returns Pixel565 values.
```
Comment (dlech) [inline fb/xrgb.go]: It looks like the At method returns color.Color, not Pixel565

File: fb/xrgb.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+type XRGB struct {
+	// Pix holds the image's pixels, as 32 bit XRGB
+	// values stored in little-endian order.
+	// The pixel at (x, y) is the four bytes at
+	// Pix[2*(x-Rect.Min.X) + (y-Rect.Min.Y)*Stride].
```
Comment (dlech) [inline fb/xrgb.go]: Again, 4 instead of 2?

File: fb/xrgb.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+	Pix []uint8
+	// Stride is the Pix stride (in bytes) between
+	// vertically adjacent pixels.
+	Stride int
+	// Rect is the image's bounds.
+	Rect image.Rectangle
+}
+
+// ColorModel returns the XRGB color model.
+func (p *XRGB) ColorModel() color.Model { return RGB565Model }
```
Comment (dlech) [inline fb/xrgb.go]: Is RGB565Model correct?

File: fb/xrgb.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+// At returns the color of the pixel at (x, y).
+func (p *XRGB) At(x, y int) color.Color {
+	if !(image.Point{x, y}.In(p.Rect)) {
+		return color.RGBA{}
+	}
+	i := p.pixOffset(x, y)
+	return PixelXRGB{R: p.Pix[i+2], G: p.Pix[i+1], B: p.Pix[i]}
+}
+
+// Set sets the color of the pixel565 at (x, y) to c.
```
Comment (dlech) [inline fb/xrgb.go]: Again, the mention of pixel565 makes me think I am missing something here.

File: fb/xrgb_test.go [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+func TestXRGB(t *testing.T) {
+	for _, test := range testImages {
+		golden := filepath.FromSlash("testdata/" + test + "-xrgb.png")
+
+		src, err := decodeImage(filepath.FromSlash("testdata/" + test + ".png"))
+		if err != nil {
+			t.Fatalf("failed to read src image file %v.png: %v", test, err)
+		}
+
+		got := NewRGB565(src.Bounds())
```
Comment (dlech) [inline fb/xrgb_test.go]: Shoudn't this be NewXRGB?

File: fb/xrgb_test.go [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+				t.Fatalf("failed to encode golden image %v-xrgb.png: %v", test, err)
+			}
+			continue
+		}
+
+		gol, err := decodeImage(golden)
+		if err != nil {
+			t.Fatalf("failed to read golden image file %v-xrgb.png: %v", test, err)
+		}
+		want := NewRGB565(gol.Bounds())
```
Comment (dlech) [inline fb/xrgb_test.go]: and also here NewXRGB?

**Comment by vladimir-ch** [comment]
Ah, sorry, I had the comments ready for some time and now I submitted them without checking the current state of discussion.

File: fb/xrgb.go (hunk 1 above)
Comment (kortschak) [reply fb/xrgb.go]: Fixed

File: fb/xrgb.go (hunk 2 above)
Comment (kortschak) [reply fb/xrgb.go]: This is fixed, but `At` has to return `color.Color` to satisfy the `image.Image` interface. The underlying concrete value is a `PixelXRGB`.

File: fb/xrgb.go (hunk 3 above)
Comment (kortschak) [reply fb/xrgb.go]: Fixed

File: fb/xrgb.go (hunk 4 above)
Comment (kortschak) [reply fb/xrgb.go]: Yes, The only difference between an `XRGB` and an `color.RGBA` is the absence of the A channel. The storage in the image's `Pix` field` is independent of the colour's representation here.

File: fb/xrgb.go (hunk 5 above)
Comment (kortschak) [reply fb/xrgb.go]: No, this is just me. Sorry.

File: fb/xrgb_test.go (hunk 6 above)
Comment (kortschak) [reply fb/xrgb_test.go]: Ugh, sorry. Fixed.

File: fb/xrgb_test.go (hunk 7 above)
Comment (kortschak) [reply fb/xrgb_test.go]: Fixed.

File: fb/xrgb.go [hunk 8]
```diff
@@ -0,0 +1,109 @@
+// Copyright ©2016 The ev3go Authors. All rights reserved.
```
[the remark this answers is not in the document]
Comment (kortschak) [reply fb/xrgb.go]: As attested by the horrific level of copy-paste errorism, the code was originally the 565 code.

File: fb/xrgb.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kortschak) [reply fb/xrgb.go]: Fixed

**Comment by kortschak** [comment]
Sorry for that. I should not code just before bed.

The issues were largely copy-paste errors. The issue about the return type is a Go-ism for interface satisfaction.

File: fb/xrgb.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (kortschak) [reply fb/xrgb.go]: Fixed

**Comment by kortschak** [comment]
The operations on the colour and alpha channels in the color model are because the [`RGBA` method of `color.Color` returns 16 bit values (in a uint32)](https://golang.org/pkg/image/color/#Color), so they need to be reduced to 8 bit width.

**Comment by kortschak** [comment]
Updates https://github.com/ev3go/ev3/issues/2.

**Pull request merged**